paper/automated.tex

```diff
@@ -1,20 +1,108 @@
+
+For saturation proofs $\ldots$ (TODO: add explanations from \url{https://leanprover.zulipchat.com/#narrow/channel/458659-Equational/topic/Vampire.20-.3E.20Lean/near/478147138})
+
+For equational proofs from external provers (e.g. MagmaEgg), we used also a simple version of reconstruction by (re-)constructing the proofs of equality from an explanation, using congruence lemmas from Lean.
+In the case of equational proofs from the \texttt{egg} tactic, these could be converted into a series of \texttt{calc} steps, documenting the explicit calculation, using the \texttt{calcify} tactic\footnote{\url{https://github.com/nomeata/lean-calcify/}}.
+
+In general, we've observed that there are multiple ways of integrating decisions procedures within Lean, with different levels of integration.
+\begin{enumerate}
+    \item Using a Lean tactic, which calls a decision procedure written in Lean (like \texttt{aesop} or \texttt{duper}).
+    \item Using a Lean tactic, which calls an existing (external) ATP and reconstructs a proof term from the ATP's result (like \texttt{bv\_decide} or \texttt{egg}).
+    \item\label{external} Using an external script which calls an existing ATP and generates a source file \texttt{.lean} which captures the result explicitly.
+\end{enumerate}
+
+
+This project primarily used the least integrated approach, (Option \ref{external}), as it was fastest and required no dependencies on the other contributors.
+This also has drawbacks: primarily, while the upfront effort is lower, the effort to use is higher than with a tactic, once the tactic is developed.
+It also makes integration with larger code bases more difficult: in this project the (mathematical) dependencies were by design very minimal, Magmas are simple, and we built our own definitions for them.
+For example, integration with the typeclass system become much more difficult when working with more complex mathematical objects that build upon many layers of structure.
+
+\paragraph{Semi-Automated Counterexample Guidance}
+
+Another use of ATPs has been in a semi-automatic fashion, to find counter-examples.
+The general strategy was to use ATPs to find counter-examples to implications by building magmas iteratively.
+If we want to build a counterexample to $l \vdash l'$, we want to build construct a magma where $l$ holds but $l'$ doesn't.
+In this method, we iteratively strengthen a construction with additional hypotheses, and use the ATP to check wether these hypotheses are not too strong (to imply $l'$) or unsound (to disallow $l$).
+
+TODO: this should also be expanded more, at least with references to some of the constructions in other chapters.
+
+While equational theories can also be used in a semi-automatic fashion to prove equations~\cite{DBLP:journals/pacmpl/KoehlerGBGTS24}, the positive implications in this project were all simple enough that we did not need a semi-automatic approach for them.
```

Comment:
equational theories -> equality saturation? While this sentence is true for the general graph, it is not completely true for either the finite implications or the Higman-Neumann work Jose Brox has done, where he has indeed performed guided search.

Reply:
fair point! I changed the formulation a bit and added a TODO to include those.